src/rt/runtime.rs

```diff
@@ -0,0 +1,449 @@
+        let cpus = num_cpus::get().max(1);
+        let processors: Vec<_> = (0..cpus).map(|_| Processor::new()).collect();
+        let stealers = processors.iter().map(|p| p.worker.stealer()).collect();
+
+        Runtime {
+            reactor: Reactor::new().unwrap(),
+            injector: Injector::new(),
+            stealers,
+            sched: Mutex::new(Scheduler {
+                processors,
+                machines: Vec::new(),
+                progress: false,
+                polling: false,
+            }),
+        }
+    }
+
+    /// Returns a reference to the reactor.
+    pub fn reactor(&self) -> &Reactor {
+        &self.reactor
+    }
+
+    /// Flushes the task slot so that tasks get run more fairly.
+    pub fn yield_now(&self) {
+        YIELD_NOW.with(|flag| flag.set(true));
+    }
+
+    /// Schedules a task.
+    pub fn schedule(&self, task: Runnable) {
+        MACHINE.with(|machine| {
+            // If the current thread is a worker thread, schedule it onto the current machine.
+            // Otherwise, push it into the global task queue.
+            match machine.get() {
+                None => {
+                    self.injector.push(task);
+                    self.notify();
+                }
+                Some(m) => m.schedule(&self, task),
+            }
+        });
+    }
+
+    /// Runs the runtime on the current thread.
+    pub fn run(&self) {
+        scope(|s| {
+            let mut idle = 0;
+            let mut delay = 0;
+
+            loop {
+                // Get a list of new machines to start, if any need to be started.
+                for m in self.make_machines() {
+                    idle = 0;
+
+                    s.builder()
+                        .name("async-std/machine".to_string())
+                        .spawn(move |_| {
+                            abort_on_panic(|| {
+                                let _ = MACHINE.with(|machine| machine.set(m.clone()));
+                                m.run(self);
+                            })
+                        })
+                        .expect("cannot start a machine thread");
+                }
+
+                // Sleep for a bit longer if the scheduler state hasn't changed in a while.
+                if idle > 10 {
+                    delay = (delay * 2).min(10_000);
+                } else {
+                    idle += 1;
+                    delay = 1000;
+                }
+
+                thread::sleep(Duration::from_micros(delay));
+            }
+        })
+        .unwrap();
+    }
+
+    /// Returns a list of machines that need to be started.
+    fn make_machines(&self) -> Vec<Arc<Machine>> {
+        let mut sched = self.sched.lock().unwrap();
+        let mut to_start = Vec::new();
+
+        // If there is a machine that is stuck on a task and not making any progress, steal its
+        // processor and set up a new machine to take over.
+        for m in &mut sched.machines {
```

Comment:
Interesting, it looks like the Runtime's run loop works in tandem with each `machine`'s run loop to set-and-detect progress for each machine and spawns things as required?

Specifically, the scheduler's Runtime's run loop invokes `make_machines`, which sets each existing Machine's progress to false, and each Machine's, while running, constantly sets it back to `true`. So long as each Machine sets its own `progress` to `true` faster than the Runtime's run loop invokes `make_machines` (dependent on the delay, which is itself dependent on the need for new machines), no new Machines get spawned?